### DISCUSSION TITLE
Discrepancy between the API & the UI for UAM fine grain authorization

### DISCUSSION THREAD

**Discussion by davidrouchet** [keycloak/keycloak#16831] [Q&A]
Hi there,

While playing around with Keycloak Authorization Services, UMA & fine grain authorization, we have been observing a discrepancy between the API & the UI.

A client has been created with Authorization enabled (fine-grained authorization support for a client). From the authorization tab, a resource has been defined as well as scopes, policies and permissions.

When the same thing is achieved using the API (i.e /admin/realms/<realm/clients endpoint, UMA resource registration endpoints and Keycloak Policy Endpoints), although the behavior of the client is the same (work as expected), when we look at the UI, we can only see the resources, but no scopes, policies nor permissions from the respective tabs.

Any idea whether this is a bug or if there is another explanation?

Tnx

**Comment by ssilvert** [comment]
Offhand, I don't know what would be different.  I would suspect that the API calls you are making are slightly different from the calls the UI is making.  The first step would be to use the browser's Inspector to look at exactly what the UI is sending, then compare to what your code is doing.

**Reply by davidrouchet** [reply]
Hi Stan,

I guess the difference is that the UI seems to use Keycloak specific APIs, one for defining the policy, the other one to define the permission (which don't seem to be documented in Keycloak, BTW) like example A below, whereas we are using a single UMA Authorization Service normalized API endpoint (that Keycloak supports) like example B below

**Example A (API called by UI)**
```
<BaseUrl>/admin/realms/<realm-name>/clients/9e074138-2b13-4057-b698-d5a40e7ec3c7/authz/resource-server/policy/client
{
  "type": "client",
  "logic": "POSITIVE",
  "decisionStrategy": "UNANIMOUS",
  "name": "Institutes read access for API Keys",
  "clients": ["8e97de83-5ab4-448b-9648-72621019c5b5"]
}
```
and
```
<BaseUrl>/admin/realms/hello-world-authz/clients/788fad29-69b3-46e6-9061-c860cbb71b6d/authz/resource-server/permission/scope/ad2bf07d-d7fc-49f3-8975-0f31d523a6a0
{
  "id": "ad2bf07d-d7fc-49f3-8975-0f31d523a6a0",
  "name": "Allow to read licenses",
  "description": "",
  "type": "scope",
  "logic": "POSITIVE",
  "decisionStrategy": "AFFIRMATIVE",
  "resources": ["f2ba907e-1ca3-4994-a0d6-9fbaf80318c9"],
  "scopes": ["23b954e9-6d7e-45f6-bf79-481c179ed392"],
  "policies": ["1d7a9fc3-4d3b-4830-86b3-c207eae97779"]
}
```

**Example B (UMA Authorization Service API)**
```
<BaseUrl>/realms/<realm-name>/authz/protection/uma-policy/:resourceId
{
        "name": "entitlement all access policy ",
        "description": "Allow to access entitlements",
        "scopes": [ 
            "read", 
            "write", 
            "create", 
            "delete"
        ],
        "logic": "POSITIVE",
        "decisionStrategy": "AFFIRMATIVE",
        "clients": [
            "api-key-2"
        ]
}
```
